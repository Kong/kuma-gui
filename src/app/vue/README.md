---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

