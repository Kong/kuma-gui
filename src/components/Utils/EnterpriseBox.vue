<template>
  <KAlert
    class="py-1"
    appearance="success"
    size="small"
  >
    <template slot="alertMessage">
      <div class="alert-content flex items-center text-sm">
        <div class="mr-3">
          Do more with Kuma <span class="alert-content__arrow">&rarr;</span>
        </div>
        <div>
          <KButton
            class="no-underline"
            appearance="primary"
            size="small"
            target="_blank"
            :to="url"
          >
            Enterprise
          </KButton>
        </div>
      </div>
    </template>
  </KAlert>
</template>

<script>
export default {
  name: 'EnterpriseBox',
  data () {
    return {
      url: 'https://kuma.io/enterprise/?utm_source=Kuma-GUI'
    }
  },
}
</script>

<style lang="scss" scoped>
.alert-content {
  display: flex;
  align-items: center;
  font-size: var(--type-sm);

  > *:first-of-type {
    margin-right: var(--spacing-sm);
  }

  .k-button {
    text-decoration: none;

    &:after {
      display: none;
    }
  }
}
</style>
