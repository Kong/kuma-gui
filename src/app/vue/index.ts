import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
