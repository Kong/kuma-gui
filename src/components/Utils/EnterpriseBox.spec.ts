import { render } from '@testing-library/vue'
import EnterpriseBox from './EnterpriseBox.vue'

describe('EnterpriseBox.vue', () => {
  it('renders snapshot', () => {
    const { container } = render(EnterpriseBox)

    expect(container).toMatchSnapshot()
  })
})
