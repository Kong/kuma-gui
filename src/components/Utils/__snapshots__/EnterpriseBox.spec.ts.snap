// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterpriseBox.vue renders snapshot 1`] = `
<div>
  <div
    class="k-alert py-1 success small"
    data-v-14c8a976=""
    role="alert"
  >
    <!---->
    <!---->
    <span
      data-v-14c8a976=""
    >
      <div
        class="alert-content flex items-center text-sm"
        data-v-14c8a976=""
      >
        <div
          class="mr-3"
          data-v-14c8a976=""
        >
          
        Do more with Kuma 
          <span
            class="alert-content__arrow"
            data-v-14c8a976=""
          >
            â†’
          </span>
        </div>
         
        <div
          data-v-14c8a976=""
        >
          <a
            class="k-button success-button small primary"
            data-v-14c8a976=""
            data-v-38fa8ecf=""
            href="https://kuma.io/enterprise/?utm_source=Kuma&utm_medium=Kuma-GUI"
            target="_blank"
            type="button"
          >
            
          Enterprise
        
          </a>
        </div>
      </div>
    </span>
  </div>
</div>
`;
